In [1]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [2]:
!pip install --upgrade pip
!pip install --upgrade setuptools wheel
!pip install --upgrade "mxnet<2.0.0"
!pip install autogluon
!pip install pydeseq2
!pip install adjustText
!pip install scanpy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 17.0 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 23.1.2
    Uninstalling pip-23.1.2:
      Successfully uninstalled pip-23.1.2
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 819.5/819.5 kB 10.7 MB/s eta 0:00:00
  Attempting uninstall: setuptools
    Found existing installation: setuptools 67.7.2
    Uninstalling setuptools-67.7.2:
      Successfully uninstalled setuptools-67.7.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16, which is not installed.
lida 0.0.10 requires fastapi, which is not installed.
lida 0.0.10 requires kaleido, which is not installed.
lida 0.0.10 requires python-multipart, which is not installed.
lida 0.0.10 requires uvicorn, which is not installed.


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.1/49.1 MB 18.0 MB/s eta 0:00:00
  Attempting uninstall: graphviz
    Found existing installation: graphviz 0.20.1
    Uninstalling graphviz-0.20.1:
      Successfully uninstalled graphviz-0.20.1
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 619.9/619.9 MB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.2/41.2 kB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.4/90.4 kB 7.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 3.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.9/116.9 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 106.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.3/79.3 kB 6.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.2/112.2 kB 9.2 MB/s 

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 kB 894.6 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.2/119.2 kB 4.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.9/90.9 kB 2.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 43.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.8/55.8 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.8/79.8 kB 7.3 MB/s eta 0:00:00
  Created wheel for umap-learn: filename=umap_learn-0.5.5-py3-none-any.whl size=86849 sha256=8d1debc0dd378d78eda42c87b8a7e59da3901aef71a2d7d6dfb93adc4e90fe45
  Stored in directory: /root/.cache/pip/wheels/3a/70/07/428d2b58660a1a3b431db59b806a10da736612ebbc66c1bcc5
  Created wheel for session-info: filename=session_info-1.0.0-py3-none-any.whl size=8047 sha256=3b31327b9b534fd9e8ce10f25fd0add2bbc9b07352acc2a470995d0232db3fba
  Stored in directo

In [5]:
import pandas as pd
from pathlib import Path

DATA = Path("/content/drive/My Drive/F23/ECBME4060/data/project")


#Mutual Exclusivity Analysis

# Binary Matrix Construction

In [47]:
import autogluon

import pandas as pd
import scanpy as sc

# Import the data
datamutations = pd.read_csv(DATA / "data_mutationslaml.txt", sep='\t')

# Get unique patients and genes
patients = datamutations['Tumor_Sample_Barcode'].unique()[0:]
numpatients = len(patients)
genes = datamutations['Hugo_Symbol'].unique()[0:]
numgenes = len(genes)

# Create a binary matrix
binarymatrix = pd.DataFrame(0, index=patients, columns=genes)

# Populate the binary matrix
for i in range(1, len(datamutations)):
    gene = datamutations.loc[i, 'Hugo_Symbol']
    patient = datamutations.loc[i, 'Tumor_Sample_Barcode']

    if gene in genes and patient in patients:
        binarymatrix.loc[patient, gene] = 1

<ipython-input-47-46eaec775d5b>:7: DtypeWarning: Columns (33,34,35,36,45) have mixed types. Specify dtype option on import or set low_memory=False.
  datamutations = pd.read_csv(DATA / "data_mutationslaml.txt", sep='\t')


In [48]:
binarymatrix.head()

TBX15  TCHHL1  DNMT3A  IDH1  GRM7  ANKRD30A  KRT74  PTPN11  \
TCGA-AB-2802-03      1       1       1     1     1         1      1       1   
TCGA-AB-2803-03      0       0       0     0     0         0      0       0   
TCGA-AB-2804-03      0       0       0     0     0         0      0       0   
TCGA-AB-2805-03      0       0       0     0     0         0      0       0   
TCGA-AB-2806-03      0       0       0     0     0         0      0       0   

                 EP400  RP11-815J4.6  ...  ACAA1  POLR2B  CDKL3  ERBB2IP  \
TCGA-AB-2802-03      1             1  ...      0       0      0        0   
TCGA-AB-2803-03      0             0  ...      0       0      0        0   
TCGA-AB-2804-03      0             0  ...      0       0      0        0   
TCGA-AB-2805-03      0             0  ...      0       0      0        0   
TCGA-AB-2806-03      0             0  ...      0       0      0        0   

                 PLAA  APOOL  PCSK2  C4orf29  CLEC7A  CALCR  
TCGA-AB-2802-03     0      0      0        0       0      0  
TCGA-AB-2803-03     0      0      0        0       0      0  
TCGA-AB-2804-03     0      0      0        0       0      0  
TCGA-AB-2805-03     0      0      0        0       0      0  
TCGA-AB-2806-03     0      0      0        0       0      0  

[5 rows x 6427 columns]

In [49]:
print(binarymatrix.shape)

(199, 6427)


There are 199 unique patients and 6427 unique mutated genes.

In [50]:
pip install numpy scipy statsmodels

# First Round of Fisher's Exact Tests

In [10]:
import numpy as np
from scipy.stats import fisher_exact
from statsmodels.stats.multitest import multipletests
import time

# Fisher's exact tests for each pair of columns
start_time = time.time()

numgenes = binarymatrix.shape[1]

# Initialize result matrices
pvals = np.zeros((numgenes, numgenes))
odds = np.zeros((numgenes, numgenes))

# Loop through every pair of genes
for i in range(numgenes):
    for j in range(numgenes):
        if i == j:
            pvals[i, j] = 1  # for the same gene, p-value is 1
        else:
            contingencytable = np.array([
                [np.sum(binarymatrix[:, i] & ~binarymatrix[:, j]), np.sum(binarymatrix[:, i] & binarymatrix[:, j])],
                [np.sum(~binarymatrix[:, i] & ~binarymatrix[:, j]), np.sum(~binarymatrix[:, i] & binarymatrix[:, j])]
            ])
            # Fisher's exact test for each pair
            oddsratio, P = fisher_exact(contingencytable, alternative='two-sided')
            # Add results to result matrices
            pvals[i, j] = P
            odds[i, j] = np.log2(oddsratio)

# Calculate FDR corrected q-values using BH method
flat_pvals = pvals.flatten()
rejected, qvals, _, _ = multipletests(flat_pvals, alpha=0.05, method='fdr_bh')

qvals = qvals.reshape((numgenes, numgenes))

end_time = time.time()
elapsed_time = end_time - start_time

print(f"Time taken: {elapsed_time} seconds")

InvalidIndexError: ignored

# First Greedy Algoritm

In [ ]:
import pandas as pd
import numpy as np

# Create DataFrame from the binary matrix and patients
binarymatrix_df = pd.DataFrame(binarymatrix, columns=genes, index=patients)

# Find the index of the "TP53" gene
tp53ind = genes.index("TP53")

# Initialize lists to store results
sigpvals = []
indices = []
me = [tp53ind]

# Find mutually exclusive genes (odds > 0)
for i in range(numgenes):
    if odds[tp53ind, i] > 0:
        sigpvals.append(pvals[tp53ind, i])
        indices.append(i)

# Find the index with the minimum p-value
ind = np.argmin(sigpvals)

# Store p-value and index
p = 1 if not sigpvals else sigpvals[ind]
pvalues = [1, p]

# Store indices
me.extend([indices[ind]])

# Display results
print(me)
print(np.array(genes)[me])
print(pvalues)

# Write results to a text file
with open('lamlresults.txt', 'w') as fid:
    fid.write(' '.join(map(str, np.array(genes)[me])) + '\n')
    fid.write(' '.join(map(str, me)) + '\n')
    fid.write(' '.join(map(str, pvalues)) + '\n\n')

# Update binary data matrix
binarymatrix_df['Module'] = binarymatrix_df.iloc[:, me].sum(axis=1)
binarymatrix_df.loc[binarymatrix_df['Module'] > 1, 'Module'] = 1
binarymatrix_df = binarymatrix_df.drop(columns=genes[me])

# Convert the DataFrame back to a NumPy array
binarymatrix = binarymatrix_df.to_numpy()

# Subsequent Fisher's Exact Tests and Greedy Algorithms

In [ ]:
import numpy as np
from scipy.stats import fisher_exact
from statsmodels.stats.multitest import multipletests
import time

# Fisher's exact tests for each pair of columns
start_time = time.time()
numgenes = binarymatrix.shape[1]

# Initialize result matrices
pvals = np.zeros((numgenes, numgenes))
odds = np.zeros((numgenes, numgenes))

# Loop through every pair of genes
for i in range(numgenes):
    for j in range(numgenes):
        if i == j:
            pvals[i, j] = 1  # for the same gene, p-value is 1
        else:
            contingencytable = np.array([
                [np.sum(binarymatrix[:, i] & ~binarymatrix[:, j]), np.sum(binarymatrix[:, i] & binarymatrix[:, j])],
                [np.sum(~binarymatrix[:, i] & ~binarymatrix[:, j]), np.sum(~binarymatrix[:, i] & binarymatrix[:, j])]
            ])
            # Fisher's exact test for each pair
            oddsratio, P = fisher_exact(contingencytable, alternative='two-sided')
            # Add results to result matrices
            pvals[i, j] = P
            odds[i, j] = np.log2(oddsratio)

# Calculate FDR corrected q-values using BH method
flat_pvals = pvals.flatten()
rejected, qvals, _, _ = multipletests(flat_pvals, alpha=0.05, method='fdr_bh')

qvals = qvals.reshape((numgenes, numgenes))

end_time = time.time()
elapsed_time = end_time - start_time

print(f"Time taken: {elapsed_time} seconds")

# Differential Expression Analysis

In [65]:
import pandas as pd

counts = pd.read_csv(DATA / "tcga_laml_raw_count.txt")

In [66]:
counts

TCGA-AB-2872-03A-01T-0735-13  TCGA-AB-2990-03B-01T-0748-13  \
BX004987.1                              0                           0.0   
AC145212.2                              0                           0.0   
AC145212.1                              0                           0.0   
Y_RNA                                   0                           0.0   
AC011043.1                              0                           0.0   
...                                   ...                           ...   
IL9R                                    0                           0.0   
WASIR1                                  0                           0.0   
AJ271736.10                             0                           0.0   
WASH6P                                  0                           0.0   
DDX11L16                                0                           0.0   

             TCGA-AB-2825-03A-01T-0736-13  TCGA-AB-2812-03A-01T-0734-13  \
BX004987.1                              0                             0   
AC145212.2                              0                             0   
AC145212.1                              0                             0   
Y_RNA                                   0                             0   
AC011043.1                              0                             0   
...                                   ...                           ...   
IL9R                                    0                             0   
WASIR1                                  0                             0   
AJ271736.10                             0                             0   
WASH6P                                  0                             0   
DDX11L16                                0                             0   

             TCGA-AB-2838-03A-01T-0736-13  TCGA-AB-2948-03A-01T-0740-13  \
BX004987.1                              0                             0   
AC145212.2                              0                             0   
AC145212.1                              0                             0   
Y_RNA                                   0                             0   
AC011043.1                              0                             0   
...                                   ...                           ...   
IL9R                                    0                             0   
WASIR1                                  0                             0   
AJ271736.10                             0                             0   
WASH6P                                  0                             0   
DDX11L16                                0                             0   

             TCGA-AB-2849-03A-01T-0734-13  TCGA-AB-2943-03A-01T-0740-13  \
BX004987.1                              0                             0   
AC145212.2                              0                             0   
AC145212.1                              0                             0   
Y_RNA                                   0                             0   
AC011043.1                              0                             0   
...                                   ...                           ...   
IL9R                                    0                             0   
WASIR1                                  0                             0   
AJ271736.10                             0                             0   
WASH6P                                  0                             0   
DDX11L16                                0                             0   

             TCGA-AB-2952-03B-01T-0760-13  TCGA-AB-2896-03B-01T-0751-13  ...  \
BX004987.1                              0                             0  ...   
AC145212.2                              0                             0  ...   
AC145212.1                              0                             0  ...   
Y_RNA                                   0                             0  ...   
AC011043.

In [67]:
countst = counts.T
countst.head()

BX004987.1  AC145212.2  AC145212.1  Y_RNA  \
TCGA-AB-2872-03A-01T-0735-13         0.0         0.0         0.0    0.0   
TCGA-AB-2990-03B-01T-0748-13         0.0         0.0         0.0    0.0   
TCGA-AB-2825-03A-01T-0736-13         0.0         0.0         0.0    0.0   
TCGA-AB-2812-03A-01T-0734-13         0.0         0.0         0.0    0.0   
TCGA-AB-2838-03A-01T-0736-13         0.0         0.0         0.0    0.0   

                              AC011043.1  AC011043.2  Y_RNA  AC011841.1  \
TCGA-AB-2872-03A-01T-0735-13         0.0         0.0    0.0         0.0   
TCGA-AB-2990-03B-01T-0748-13         0.0         0.0    0.0         0.0   
TCGA-AB-2825-03A-01T-0736-13         0.0         0.0    0.0         0.0   
TCGA-AB-2812-03A-01T-0734-13         0.0         0.0    0.0         0.0   
TCGA-AB-2838-03A-01T-0736-13         0.0         0.0    0.0         0.0   

                              KIR3DL2  KIR2DS1  ...  AMD1P2  DPH3P2  VAMP7  \
TCGA-AB-2872-03A-01T-0735-13      0.0      0.0  ...     0.0     0.0    0.0   
TCGA-AB-2990-03B-01T-0748-13      0.0      0.0  ...     0.0     0.0    0.0   
TCGA-AB-2825-03A-01T-0736-13      0.0      0.0  ...     0.0     0.0    0.0   
TCGA-AB-2812-03A-01T-0734-13      0.0      0.0  ...     0.0     0.0    0.0   
TCGA-AB-2838-03A-01T-0736-13      0.0      0.0  ...     0.0     0.0    0.0   

                              ELOCP24  TRPC6P  IL9R  WASIR1  AJ271736.10  \
TCGA-AB-2872-03A-01T-0735-13      0.0     0.0   0.0     0.0          0.0   
TCGA-AB-2990-03B-01T-0748-13      0.0     0.0   0.0     0.0          0.0   
TCGA-AB-2825-03A-01T-0736-13      0.0     0.0   0.0     0.0          0.0   
TCGA-AB-2812-03A-01T-0734-13      0.0     0.0   0.0     0.0          0.0   
TCGA-AB-2838-03A-01T-0736-13      0.0     0.0   0.0     0.0          0.0   

                              WASH6P  DDX11L16  
TCGA-AB-2872-03A-01T-0735-13     0.0       0.0  
TCGA-AB-2990-03B-01T-0748-13     0.0       0.0  
TCGA-AB-2825-03A-01T-0736-13     0.0       0.0  
TCGA-AB-2812-03A-01T-0734-13     0.0       0.0  
TCGA-AB-2838-03A-01T-0736-13     0.0       0.0  

[5 rows x 63856 columns]

In [15]:
!pip install rpy2

In [68]:
import os
import pickle as pkl

from pydeseq2.dds import DeseqDataSet
from pydeseq2.ds import DeseqStats
from pydeseq2.utils import load_example_data

print("Binary Matrix Index:", binarymatrix.index)
print("Counts Index:", countst.index)

common_samples = binarymatrix.index.intersection(countst.index)
binarymatrix_commonsamps = binarymatrix.loc[common_samples, :]
rawcounts = countst.loc[common_samples, :]
binarymatrix_tp53=binarymatrix['TP53']
countst_index_subset = [index[:15] for index in countst.index.values]

# Filter binarymatrix_tp53 based on the index in countst
countst = countst[countst.index.str[:15].isin(binarymatrix_tp53.index)]
binarymatrix_tp53_filt = binarymatrix_tp53[binarymatrix_tp53.index.isin(countst_index_subset)]
list_data = {'Condition':[ 'Mutated TP53' if i == 1 else "Wild-Type TP53" for i in binarymatrix_tp53_filt],
                  'Patient':  [ i for i in binarymatrix_tp53_filt.index]} #fix
df_data = pd.DataFrame(list_data, index=countst.index.values.tolist()) #fix
df_data.head()

Binary Matrix Index: Index(['TCGA-AB-2802-03', 'TCGA-AB-2803-03', 'TCGA-AB-2804-03',
       'TCGA-AB-2805-03', 'TCGA-AB-2806-03', 'TCGA-AB-2807-03',
       'TCGA-AB-2808-03', 'TCGA-AB-2809-03', 'TCGA-AB-2810-03',
       'TCGA-AB-2811-03',
       ...
       'TCGA-AB-3002-03', 'TCGA-AB-3005-03', 'TCGA-AB-3006-03',
       'TCGA-AB-3007-03', 'TCGA-AB-3008-03', 'TCGA-AB-3009-03',
       'TCGA-AB-3011-03', 'TCGA-AB-3012-03', 'TCGA-AB-2815-03',
       'TCGA-AB-2944-03'],
      dtype='object', length=199)
Counts Index: Index(['TCGA-AB-2872-03A-01T-0735-13', 'TCGA-AB-2990-03B-01T-0748-13',
       'TCGA-AB-2825-03A-01T-0736-13', 'TCGA-AB-2812-03A-01T-0734-13',
       'TCGA-AB-2838-03A-01T-0736-13', 'TCGA-AB-2948-03A-01T-0740-13',
       'TCGA-AB-2849-03A-01T-0734-13', 'TCGA-AB-2943-03A-01T-0740-13',
       'TCGA-AB-2952-03B-01T-0760-13', 'TCGA-AB-2896-03B-01T-0751-13',
       ...
       'TCGA-AB-2921-03A-01T-0740-13', 'TCGA-AB-3011-03A-01T-0736-13',
       'TCGA-AB-2885-03A-01T-0735-13', 'TCGA-A

Condition          Patient
TCGA-AB-2872-03A-01T-0735-13  Wild-Type TP53  TCGA-AB-2803-03
TCGA-AB-2990-03B-01T-0748-13  Wild-Type TP53  TCGA-AB-2805-03
TCGA-AB-2825-03A-01T-0736-13  Wild-Type TP53  TCGA-AB-2806-03
TCGA-AB-2812-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2807-03
TCGA-AB-2838-03A-01T-0736-13  Wild-Type TP53  TCGA-AB-2808-03

In [55]:
df_data.head(20)

Condition          Patient
TCGA-AB-2872-03A-01T-0735-13  Wild-Type TP53  TCGA-AB-2803-03
TCGA-AB-2990-03B-01T-0748-13  Wild-Type TP53  TCGA-AB-2805-03
TCGA-AB-2825-03A-01T-0736-13  Wild-Type TP53  TCGA-AB-2806-03
TCGA-AB-2812-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2807-03
TCGA-AB-2838-03A-01T-0736-13  Wild-Type TP53  TCGA-AB-2808-03
TCGA-AB-2948-03A-01T-0740-13  Wild-Type TP53  TCGA-AB-2810-03
TCGA-AB-2849-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2811-03
TCGA-AB-2943-03A-01T-0740-13  Wild-Type TP53  TCGA-AB-2812-03
TCGA-AB-2952-03B-01T-0760-13    Mutated TP53  TCGA-AB-2813-03
TCGA-AB-2896-03B-01T-0751-13  Wild-Type TP53  TCGA-AB-2814-03
TCGA-AB-2893-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2816-03
TCGA-AB-2987-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2817-03
TCGA-AB-2949-03B-01T-0748-13  Wild-Type TP53  TCGA-AB-2818-03
TCGA-AB-2878-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2819-03
TCGA-AB-2881-03A-01T-0735-13    Mutated TP53  TCGA-AB-2820-03
TCGA-AB-2875-03A-01T-0735-13  Wild-Type TP53  TCGA-AB-2821-03
TCGA-AB-2996-03A-01T-0735-13  Wild-Type TP53  TCGA-AB-2822-03
TCGA-AB-2931-03A-01T-0740-13  Wild-Type TP53  TCGA-AB-2823-03
TCGA-AB-2892-03A-01T-0734-13  Wild-Type TP53  TCGA-AB-2824-03
TCGA-AB-2898-03A-01T-0735-13  Wild-Type TP53  TCGA-AB-2825-03

In [69]:
duplicate_indices = countst[countst.index.duplicated(keep=False)].index
print("Duplicate Indices:", duplicate_indices)
# Option 1: Drop duplicates
countst = countst[~countst.index.duplicated(keep='first')]

# Option 2: Aggregate duplicates (if applicable)
# countst = countst.groupby(countst.index).sum()

# Option 3: Reset index to create a new unique index
#countst = countst.reset_index(drop=True)
countst_filtered = countst[countst.sum(axis=1) >= 1]

Duplicate Indices: Index([], dtype='object')


In [78]:
df_data = df_data[~df_data.index.duplicated(keep='first')]
countst_filtered = countst_filtered[~countst_filtered.index.duplicated(keep='first')]
countst_filtered = countst_filtered.loc[:, ~countst_filtered.columns.duplicated()]


In [71]:
df_data.shape


(177, 2)

In [79]:
countst_filtered.shape

(177, 56937)

In [60]:
df_data.index.nunique

<bound method IndexOpsMixin.nunique of Index(['TCGA-AB-2872-03A-01T-0735-13', 'TCGA-AB-2990-03B-01T-0748-13',
       'TCGA-AB-2825-03A-01T-0736-13', 'TCGA-AB-2812-03A-01T-0734-13',
       'TCGA-AB-2838-03A-01T-0736-13', 'TCGA-AB-2948-03A-01T-0740-13',
       'TCGA-AB-2849-03A-01T-0734-13', 'TCGA-AB-2943-03A-01T-0740-13',
       'TCGA-AB-2952-03B-01T-0760-13', 'TCGA-AB-2896-03B-01T-0751-13',
       ...
       'TCGA-AB-2921-03A-01T-0740-13', 'TCGA-AB-3011-03A-01T-0736-13',
       'TCGA-AB-2885-03A-01T-0735-13', 'TCGA-AB-2985-03A-01T-0734-13',
       'TCGA-AB-2929-03A-01T-0735-13', 'TCGA-AB-2978-03A-01T-0734-13',
       'TCGA-AB-2821-03A-01T-0735-13', 'TCGA-AB-2897-03A-01T-0735-13',
       'TCGA-AB-2994-03A-01T-0735-13', 'TCGA-AB-2975-03A-01T-0734-13'],
      dtype='object', length=177)>

In [80]:
dds = DeseqDataSet(
    counts=countst_filtered,
    metadata=df_data,
    design_factors="Condition",
    refit_cooks=True,
   # n_cpus=1,
)

In [81]:
dds.obs_names_make_unique


<bound method AnnData.obs_names_make_unique of AnnData object with n_obs × n_vars = 177 × 56937
    obs: 'Condition', 'Patient'
    obsm: 'design_matrix'>

In [82]:
dds.var_names_make_unique

<bound method AnnData.var_names_make_unique of AnnData object with n_obs × n_vars = 177 × 56937
    obs: 'Condition', 'Patient'
    obsm: 'design_matrix'>

In [83]:
dds.deseq2()

Fitting size factors...
... done in 0.76 seconds.

Fitting dispersions...
... done in 126.20 seconds.

Fitting dispersion trend curve...
... done in 4.55 seconds.

Fitting MAP dispersions...
... done in 137.17 seconds.

Fitting LFCs...
... done in 118.75 seconds.

Refitting 15711 outliers.

Fitting dispersions...
... done in 26.91 seconds.

Fitting MAP dispersions...
... done in 23.47 seconds.

Fitting LFCs...
... done in 47.19 seconds.

/usr/local/lib/python3.10/dist-packages/pydeseq2/dds.py:1019: ImplicitModificationWarning: Setting element `.varm['_normed_means']` of view, initializing view as actual.
  self[:, self.new_all_zeroes_genes].varm["_normed_means"] = np.zeros(
/usr/local/lib/python3.10/dist-packages/pydeseq2/dds.py:1022: ImplicitModificationWarning: Setting element `.varm['LFC']` of view, initializing view as actual.
  self[:, self.new_all_zeroes_genes].varm["LFC"] = np.zeros(


In [84]:
dds.calculate_cooks()
dds.refit()

Refitting 15711 outliers.

Fitting dispersions...
... done in 27.93 seconds.

Fitting MAP dispersions...
... done in 22.84 seconds.

Fitting LFCs...
... done in 47.42 seconds.

/usr/local/lib/python3.10/dist-packages/pydeseq2/dds.py:1019: ImplicitModificationWarning: Setting element `.varm['_normed_means']` of view, initializing view as actual.
  self[:, self.new_all_zeroes_genes].varm["_normed_means"] = np.zeros(
/usr/local/lib/python3.10/dist-packages/pydeseq2/dds.py:1022: ImplicitModificationWarning: Setting element `.varm['LFC']` of view, initializing view as actual.
  self[:, self.new_all_zeroes_genes].varm["LFC"] = np.zeros(


In [85]:
stat_res = DeseqStats(dds, alpha=0.05)
stat_res.run_wald_test()
stat_res._p_value_adjustment()

Running Wald tests...
... done in 20.47 seconds.



In [86]:
df_results=stat_res.summary()

Log2 fold change & Wald test p-value: Condition Wild-Type TP53 vs Mutated TP53
              baseMean  log2FoldChange     lfcSE      stat    pvalue  padj
BX004987.1    0.000000             NaN       NaN       NaN       NaN   NaN
AC145212.2    0.000000             NaN       NaN       NaN       NaN   NaN
AC145212.1    0.000000             NaN       NaN       NaN       NaN   NaN
Y_RNA         0.000000             NaN       NaN       NaN       NaN   NaN
AC011043.1    0.000000             NaN       NaN       NaN       NaN   NaN
...                ...             ...       ...       ...       ...   ...
CYCSP49       0.275554        0.939597  5.256789  0.178740  0.858142   1.0
TPTE2P4      69.188551        0.148440  0.770373  0.192686  0.847205   1.0
SLC25A15P1   16.287079        0.215332  2.224455  0.096802  0.922884   1.0
FAM58CP     137.030100        0.509702  0.750744  0.678929  0.497183   1.0
PARP4P1     518.561871       -0.072951  0.619390 -0.117779  0.906243   1.0

[56937 rows x 6 colu

In [87]:
stat_res.LFC

intercept  Condition_Wild-Type TP53_vs_Mutated TP53
BX004987.1        NaN                                       NaN
AC145212.2        NaN                                       NaN
AC145212.1        NaN                                       NaN
Y_RNA             NaN                                       NaN
AC011043.1        NaN                                       NaN
...               ...                                       ...
CYCSP49     -1.688742                                  0.651279
TPTE2P4      4.142106                                  0.102891
SLC25A15P1   2.652599                                  0.149256
FAM58CP      4.592355                                  0.353298
PARP4P1      6.297216                                 -0.050566

[56937 rows x 2 columns]

In [89]:
res_pdDF = pd.DataFrame({'lfc':stat_res.LFC['Condition_Wild-Type TP53_vs_Mutated TP53'],
                      'padj':stat_res.padj
                      })
res_pdDF.index = stat_res.LFC.index
sorted_padj = res_pdDF.sort_values(by='padj')
sorted_padj.head(10)

lfc          padj
CTD-2006M22.2  30.000000  3.376491e-24
RNA5SP233      30.000000  8.162843e-22
SMPDL3A         2.544378  7.778663e-05
LDLRAD3         2.118910  3.604674e-04
STK33           2.225901  6.689357e-04
PKP2            2.740740  7.839514e-04
MMP8            2.585068  9.957684e-04
OR7D2           2.317245  1.131677e-03
GABRE           3.454461  1.166596e-03
MEG3            2.944232  1.225007e-03

In [90]:
sig_genes_WT=sorted_padj[sorted_padj["lfc"]<0]
sig_genes_WT.index[:10]

Index(['PCSK6', 'CHML', 'RP11-299G20.2', 'ABCA6', 'FGF7', 'SLC38A4', 'BMS1P10',
       'CACNA1D', 'RP11-158G18.1', 'RP11-474P2.6'],
      dtype='object')

In [91]:
sig_genes_Mut=sorted_padj[sorted_padj["lfc"]>0]
sig_genes_Mut.index[:10]

Index(['CTD-2006M22.2', 'RNA5SP233', 'SMPDL3A', 'LDLRAD3', 'STK33', 'PKP2',
       'MMP8', 'OR7D2', 'GABRE', 'MEG3'],
      dtype='object')

In [92]:
sig_genes = sorted_padj[sorted_padj['padj']<=0.05]
textfile = 'sig_genes.txt'
sig_genes.to_csv(textfile, sep='\t', index=False)
sig_genes

lfc          padj
CTD-2006M22.2  30.000000  3.376491e-24
RNA5SP233      30.000000  8.162843e-22
SMPDL3A         2.544378  7.778663e-05
LDLRAD3         2.118910  3.604674e-04
STK33           2.225901  6.689357e-04
...                  ...           ...
RN7SL441P       0.558483  4.633311e-02
OSBPL9P4        4.935610  4.942349e-02
RHBDL3          1.425166  4.969482e-02
FGF13           2.459835  4.969482e-02
BPI             1.573223  4.979097e-02

[285 rows x 2 columns]